mdn/developer-portal
# 1143 + 1144: Task completion survey (core implementation)

## DO NOT MERGE TO MASTER YET - PLEASE REVIEW, BUT  THIS REQUIRES A SURVEY TO BE SET UP BEFORE PRODUCTION

This changeset adds an initial/MVP version of a prompt that asks the user to complete task-completion survey

To achieve this we:

* show a notification bar
* to 5% of visitors (configurable via settings)
* including a link to a new page (configurable via settings)
* which shows a survey the user can fill in to tell us about their experience

It uses client-side JS to handle selecting the 5% of users and setting a cookie to remember that decision, as well as whether the user has chosen to close the notification. (We looked at using django-waffle to control its appearance trigger, as per MDN, but found we had to make too many compromises to avoid poisoning the CDN.)

Re the CDN: note that at the moment, it's still set to `Vary: Accept-Encoding,Cookie`, which isn't great given that the client-side JS sets a cookie that then computes a new CDN cache key for another cached response that's identical to what we get back without that cookie in a CDN cache key. However, remedying that is ticketed as #1167.

Also note that the NotificationBar component from Mozilla Protocol is currently rendered into the base HTML as a `hidden` node, and then revealed via JS. It can be triggered entirely from JS, but until my [enhancement PR](https://github.com/mozilla/protocol/pull/534) is merged, it's not possible to add a hyperlink to the notification's content via JS. So pre-rendering was the way forward.

(Related issues #1143 and #1144)

## How to test

- check out the site locally or - **ideally** - use the [dev server via its CDN](https://developer-portal-cdn.dev.mdn.mozit.cloud/) 
- if local, in your `local.py` set `TASK_COMPLETION_SURVEY_URL` to something, eg: `"https://example.com/task-completion-survey"`  and `TASK_COMPLETION_SURVEY_PERCENTAGE` to `Decimal("99.9")`
- if on the dev server, note that it's been configured to be shown to 95% of people, and once you get shown it or not shown it, it will stay that way because a cookie is set. Feel free to zap your cookies.
- load the homepage (or any site page) and you will see the notification bar, with a link to the test URL you configured above. This will open a new tab/window for the URL if you click it
- clicking the close `X` on the notification will dismisss it (and it will not be re-shown unless you edit the `devportal_show_task_completion_survey` cookie to be `true`, or you wipe that cookie)

Cookie on load, triggering the message to be shown:
![Screenshot 2020-02-05 at 14 03 26](https://user-images.githubusercontent.com/101457/73890145-34091d80-4869-11ea-96e5-0cd22901136c.png)

Cookie after notification bar has been dismissed:
![Screenshot 2020-02-05 at 14 03 38](https://user-images.githubusercontent.com/101457/73890147-353a4a80-4869-11ea-9ae0-e40f60478e50.png)

(Try this for yourself!)

### Example UI screenshots

<img width="1013" alt="Screenshot 2020-01-28 at 17 16 21" src="https://user-images.githubusercontent.com/101457/73287940-f9acda00-41f1-11ea-9419-6f40ec94bd9c.png">
<img width="371" alt="Screenshot 2020-01-28 at 17 16 30" src="https://user-images.githubusercontent.com/101457/73287941-f9acda00-41f1-11ea-8211-bdc6ba252fc8.png">
